rfcs/converged-implementation-patterns.md

```diff
@@ -0,0 +1,171 @@
+
+Renders the correct JSX output of the component and its slots given the correct state.
+
+```typescript=
+export const renderSample = (state: SampleState) => {
+  const { slots, slotProps } = getSlots(state);
+
+  return (
+    <slots.root {...slotProps.root}>
+      <slot.otherSlot {...slotProps.otherSlot} />
+    </slots.root>
+  );
+};
+```
+
+#### useSample.ts
+
+Accepts the component props and handles and internal state that the component might need.
+
+`State` here can be pretty broad, you could also consume context or create effects. This hook should be what the component relies on to function/render
+
+```typescript=
+/**
+ * Defines the different slots that can be rendered in this component
+ *
+ * This means that an icon can be rendered 'somewhere'
+ */
+export const sampleShorthandProps = ['icon'];
+
+// Creates a helper function to merge props while respecting slot props
+const mergeProps = makeMergeProps({ deepMerge: sampleShorthandProps });
+
+/**
+ * @parameter props -> these are normal React props for JSX components
+ * @parameter ref -> In case someone wants a ref to the root DOM element
+ * @parameter defaultProps -> safe default props
+ */
+export const useSample = (props: SampleProps, ref: React.Ref<HTMLElement>, defaultProps?: SampleProps) => {
+  const resolvedRef = ref || React.useRef();
+  const { checkedValues, onCheckedValuesChange } = useMenuListContext();
+
+  // merges the props we declare internally and what is passed in
+  // by a consumer
+  const mergedProps = mergeProps(
+    {
+      ref: resolvedRef,
+      as: 'div',
+      icon: { as: 'span' },
+    },
+    defaultProps,
+    resolveShorthandProps(props, sampleShorthandProps),
+  );
+
+  if (checkedValues || onCheckedValuesChange) {
+    mergedProps.hasCheckMark = true;
+  }
+
+  // Anything else the component needs to manage its own state and render
+  [someState, setSomeState] = React.useState();
+  const { contextValue } = React.useContext();
+  React.useEffect(...);
+
+  // make an 'uber' state
+  const state = { someState, contextValue, ...mergedProps}
+
+  return state;
+};
+```
+
+#### useSampleClasses/useSampleStyles.ts
+
+Hook that accepts state and applies classnames to props and any shorthand slot props to style the component and its slots
+
+#### Sample.tsx
+
+The 'final' product, simply forwards a ref and uses all the previous building blocks to expose the final component
+
+```typescript=
+export const Sample = React.forwardRef<HTMLElement, SampleProps>((props, ref) => {
+  const state = useSample(props, ref);
+  useSampleStyles(state);
+
+  return renderSample(state);
+});
+```
+
+#### Sample.types.ts
+
+Utility types stored here
+
+### Pros and Cons
+
+#### Pros
+
+Since there is a separation of state, render, styling into different hooks/utility functions, it can be very easy for consumers to create their own custom component variants without needing to approach the library team and accommodate their custom requirements
```

Comment:
```suggestion
Since there is a separation of state, render, styling into different hooks/utility functions, it can be very easy for consumers to create their own custom component variants without needing to approach the library team and accommodate their custom requirements.
```

Reply:
This is great! I can already see another guide on how that can be accomplished, but one step at a time.